hep/proposals/0000-decorators.md

```diff
@@ -0,0 +1,299 @@
+    # We will need to use something like python-varname
+    # to automatically create a task called "setup_task" here.
+    setup_task = setup()
+    # Note how easy it is to reference variables from the DAG template input
+    # or previous tasks.
+    task_a = concat(ConcatInput(word_a=worker_input.value_a, word_b=setup_task.result))
+    task_b = concat(ConcatInput(word_a=worker_input.value_b, word_b=setup_task.result))
+    final_task = concat(ConcatInput(word_a=task_a.result, word_b=task_b.result))
+    # Note, we will automatically construct the dependency graph
+    # based on the input/output relationship between tasks.
+    # Users will still be able to explicitly add dependencies between tasks
+    # using the rshift operator if they wish to define dependencies amongst
+    # tasks that don't share variables directly
+    return WorkerOutput(value=final_task.result)
+```
+
+
+## Steps Template
+
+```python
+from hera.workflows import WorkflowTemplate
+from hera import io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the a template input
+class CalculatorInput(hio.Input):
+    x: int
+    y: int
+    operation: Literal["add", "sub"] = "add"
+
+@wt.script  # Adds a new script template to the workflow template called calculator
+def calculator(calc_input: CalculatorInput) -> hio.Output:
+    if calc_input.operation == "add":
+        return hio.Output(calc_input.x + calc_input.y)
+    return hio.Output(calc_input.x - calc_input.y)
+    
+
+# This defines the another template input
+class FiboInput(hio.Input):
+    num: int
+
+class FiboOutput(hio.Output):
+    num: int
+
+@wt.entrypoint
+@wt.steps  # Adds a new script template to the workflow template called fibonacci
+def fibonacci(fibo: FiboInput) -> FiboOutput:
+    # we will need to validate that users can only run/define certain kinds of expressions here
+    # and throw warnings if they are using syntax that is invalid
+    valid_number = hio.expr(fibo.num > 1)
+    previous_num = calculator(CalculatorInput(x=fibo.num, y=1, operation="sub"), when=valid_number)
+    previous_fibo = fibonacci(FiboInput(num=previous_num.result), when=valid_number)
+    current = calculator(CalculatorInput(x=fibo.num, y=previous_fibo.num), when=valid_number)
+    return FiboOutput(num=valid_number.check(current.num, 1))
+```
+
+
+
+## Container Template
+
+```python
+from hera.workflows import WorkflowTemplate, Container
+from hera import io as hio
+
+# We start by defining our Workflow Template
+wt = WorkflowTemplate(name="my-template")
+
+# This defines the a template input
+class MyInput(hio.Input):
+    user: str = "Hera"
+
+class MyOutput(hio.Output):
+    container_greeting: Annotated[str, Parameter(name="container-greeting")]
+
+@wt.entrypoint
+@wt.container(command=["sh", "-c"], args=["echo {{input.parameters.user}}"])
+def basic_hello_world(my_input: MyInput) -> hio.Output:
```

Comment:
Just to make sure the example is complete - looking at the `container-greeting` param, looks like it needs a `value_from`, and the command should be writing to that path (as we don't have the auto-IO stuff from the runner) - using `tee` as seen in upstream examples:

```suggestion
class MyOutput(hio.Output):
    container_greeting: Annotated[
        str,
        Parameter(
            name="container-greeting",
            value_from={"path": "/tmp/hello_world.txt"},
        ),
    ]

@wt.entrypoint
@wt.container(command=["sh", "-c"], args=["echo {{input.parameters.user}} | tee /tmp/hello_world.txt"])
def basic_hello_world(my_input: MyInput) -> hio.Output:
```

Adding the whole
```py
        Parameter(
            name="container-greeting",
            value_from={"path": "/tmp/hello_world.txt"},
        ),
```
looks to be annoying though